### PR TITLE
Restrict contents of array and struct

### LINKED ISSUE #1409: WGSL: limit types that can go into arrays and structs
We don't want a pointer, texture, or sampler to exist in a structure or array.  (In the future we'll want arrays of textures.)

Currently that's not captured.

Because of the recursive nature of these types, we can't just define a new class of types similar to host-shareable or IO-shareable. That would introduce circularity.

### PR DESCRIPTION
**Pull request by dneto0**
Fixes #1409

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -347,6 +347,13 @@ A vector type is a <dfn dfn>numeric vector</dfn> type if its component type is a
                        These may only appear in specific contexts.<br>
 </table>
 
+An array element type must not be:
+
+* The [=void=] type.
```
Comment (dneto0) [inline wgsl/index.bs:352]: This is a bit of a laundry list.

I *could* have defined a new concept such as a "concrete" type.

However, I expect WGSL will soon have to support arrays of textures and arrays of samplers, so the category isn't going to be particularly stable.

File: wgsl/index.bs [hunk 2]
```diff
@@ -347,6 +347,13 @@ A vector type is a <dfn dfn>numeric vector</dfn> type if its component type is a
                        These may only appear in specific contexts.<br>
 </table>
 
+An array element type must be one of:
```
Comment (kvark) [inline wgsl/index.bs:350]: I assume this set of types is going to show up a lot. I.e. the same should apply to structure members, function arguments, etc. Can we have a common definition for this?

Comment (dneto0) [reply wgsl/index.bs:350]: We have two cases in this PR.

The closest analog in SPIR-V is "concrete" type.  (That also covers a physical pointer, which is not in WSGL).   I think we should keep an eye on this and if we get a lot more cases then we'll add a definition.

**Comment by dneto0** [comment]
Rebased to avoid merge conflicts.

**Comment by dneto0** [comment]
The change was approved in the WGSL meeting of [2021-02-16](https://docs.google.com/document/d/11bjujbElwcQYQjJFCFJhy4jgwfUG5AWeNwsDRpnQbCU/edit)

**Review by dj2 — APPROVED** [review]

**Pull request merged**